### PR TITLE
Bratseth/deprovisioned node state

### PR DESCRIPTION
**Pull request by bratseth**

### CODE REVIEW

File: node-repository/src/main/java/com/yahoo/vespa/hosted/provision/NodeRepository.java [hunk 1]
```diff
@@ -608,8 +608,11 @@ public Node markNodeAvailableForNewAllocation(String hostname, Agent agent, Stri
                 List<Node> children = list().childrenOf(node).asList();
                 children.forEach(child -> requireRemovable(child, true, force));
                 db.removeNodes(children);
-                move(node, State.deprovisioned, Agent.system, Optional.empty());
                 List<Node> removed = new ArrayList<>(children);
+                if (zone.cloud().value().equals("aws"))
```
Comment (mpolden) [inline node-repository/src/main/java/com/yahoo/vespa/hosted/provision/NodeRepository.java]: Looks like we have this in several places now. We should find a way to model it properly soon.

File: node-admin/src/main/java/com/yahoo/vespa/hosted/node/admin/configserver/noderepository/NodeState.java [hunk 2]
```diff
@@ -9,5 +9,5 @@
  * @author freva
  */
 public enum NodeState {
-    provisioned, ready, reserved, active, inactive, dirty, failed, parked
+    provisioned, ready, reserved, active, inactive, dirty, failed, parked, deprovisioned
```
Comment (mpolden) [inline node-admin/src/main/java/com/yahoo/vespa/hosted/node/admin/configserver/noderepository/NodeState.java]: Should also be added to `com.yahoo.vespa.hosted.controller.api.integration.noderepository.NodeState`.

File: config-provisioning/src/main/java/com/yahoo/config/provision/NodeType.java [hunk 3]
```diff
@@ -45,7 +45,7 @@
         this.description = description;
     }
 
-    public boolean isHost() {
+    public boolean isDockerHost() {
```
Comment (mpolden) [inline config-provisioning/src/main/java/com/yahoo/config/provision/NodeType.java]: Why not?

File: node-repository/src/main/java/com/yahoo/vespa/hosted/provision/NodeRepository.java (hunk 1 above)
Comment (bratseth) [reply node-repository/src/main/java/com/yahoo/vespa/hosted/provision/NodeRepository.java]: Agree. What I want to do is create a "cloud config" which defines the properties of a cloud and then only test for those properties.

File: config-provisioning/src/main/java/com/yahoo/config/provision/NodeType.java (hunk 3 above)
Comment (bratseth) [reply config-provisioning/src/main/java/com/yahoo/config/provision/NodeType.java]: Breaks the config server/config model API - in theory, did not look at usages.
I still want to do it but in a separate PR.

File: node-admin/src/main/java/com/yahoo/vespa/hosted/node/admin/configserver/noderepository/NodeState.java (hunk 2 above)
Comment (bratseth) [reply node-admin/src/main/java/com/yahoo/vespa/hosted/node/admin/configserver/noderepository/NodeState.java]: Done - thanks

**Review by mpolden — APPROVED** [review]
Great!

File: node-repository/src/main/java/com/yahoo/vespa/hosted/provision/node/Agent.java [hunk 4]
```diff
@@ -17,5 +17,6 @@
     FailedExpirer,
     InactiveExpirer,
     ProvisionedExpirer,
-    ReservationExpirer
+    ReservationExpirer,
+    DynamicProvisioningMaintainer
```
Comment (freva) [inline node-repository/src/main/java/com/yahoo/vespa/hosted/provision/node/Agent.java]: Add to https://github.com/vespa-engine/vespa/blob/ddd5141d112d44f881da17e83d26ac18ad35c3c1/controller-api/src/main/java/com/yahoo/vespa/hosted/controller/api/integration/noderepository/NodeHistory.java#L35

Comment (bratseth) [reply node-repository/src/main/java/com/yahoo/vespa/hosted/provision/node/Agent.java]: Thanks 

https://github.com/vespa-engine/vespa/pull/12523

**Pull request merged**